File: core/src/ionium/aabbcollision/PhysicsBody.java
package ionium.aabbcollision;

import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Pool.Poolable;
import ionium.util.MathHelper;

public class PhysicsBody implements Poolable {

	public static final PhysicsBody tmp = new PhysicsBody();
	public static final PhysicsBody tmp2 = new PhysicsBody();
	private static final Rectangle pathHitTempRect = new Rectangle();

	public Rectangle bounds = new Rectangle();
	public Vector2 velocity = new Vector2();

	public PhysicsBody() {
		this(0, 0, 1, 1);
	}

	public PhysicsBody(float x, float y, float w, float h) {
		setBounds(x, y, w, h);
	}

	public PhysicsBody setBounds(float x, float y, float w, float h) {
		bounds.set(x, y, w, h);

		return this;
	}

	public PhysicsBody setVelocity(float x, float y) {
		velocity.set(x, y);

		return this;
	}

	public boolean mayBeHitInPath(float timeScale, PhysicsBody other) {
		pathHitTempRect.set(other.getAreaOfTravel(timeScale));

		return bounds.overlaps(pathHitTempRect) || pathHitTempRect.overlaps(bounds);
	}

	public Rectangle getAreaOfTravel(float timeScale) {
		pathHitTempRect.set(bounds.x, bounds.y, velocity.x * timeScale + bounds.width,
				velocity.y * timeScale + bounds.height);
		MathHelper.normalizeRectangle(pathHitTempRect);

		return pathHitTempRect;
	}

	@Override
	public String toString() {
		return "[" + bounds.toString() + ", " + velocity.toString() + "]";
	}

	@Override
	public void reset() {
		setVelocity(0, 0);
	}

}

File: core/src/main/kotlin/io/github/chrislo27/toolboks/aabbcollision/PhysicsBody.kt
package io.github.chrislo27.toolboks.aabbcollision

import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.utils.Pool


/**
 * A simple object with a bounds rectangle and a velocity.
 */
class PhysicsBody() : Pool.Poolable {

    val bounds = Rectangle(0f, 0f, 1f, 1f)
    val velocity = Vector2()
    var metadata: Any = Unit

    constructor(x: Float, y: Float, width: Float, height: Float, vx: Float, vy: Float) : this() {
        bounds.set(x, y, width, height)
        velocity.set(vx, vy)
    }

    override fun reset() {
        bounds.set(0f, 0f, 1f, 1f)
        velocity.set(0f, 0f)
        metadata = Unit
    }
}